lib/client/conntest/database.go

```diff
@@ -0,0 +1,394 @@
+	databaseServiceName := databaseServer.GetName()
+
+	var databasePinger DatabasePinger
+
+	switch databaseProtocol {
+	case defaults.ProtocolPostgres:
+		databasePinger = database.PostgresPinger{}
+	default:
+		return nil, trace.Errorf("Database protocol not supported yet for Test Connection")
+	}
+
+	if _, err := s.appendDiagnosticTrace(ctx,
+		connectionDiagnosticID,
+		types.ConnectionDiagnosticTrace_RBAC_DATABASE,
+		"A Database Agent is available to proxy the connection to the Database.",
+		nil,
+	); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	list, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	proxyMiddleware := client.NewALPNCertChecker().
+		WithIdentityCheckerFunc(func(identity *tlsca.Identity) error {
+			if req.DatabaseUser != identity.RouteToDatabase.Username {
+				msg := fmt.Sprintf("certificate subject is for user %s, but need %s", identity.RouteToDatabase.Username, req.DatabaseUser)
+				return trace.Wrap(errors.New(msg))
+			}
+
+			if req.DatabaseName != identity.RouteToDatabase.Database {
+				msg := fmt.Sprintf("certificate subject is for database name %s, but need %s", identity.RouteToDatabase.Database, req.DatabaseName)
+				return trace.Wrap(errors.New(msg))
+			}
+
+			return nil
+		})
+
+	err = client.RunALPNAuthTunnel(ctx, client.RunALPNAuthTunnelRequest{
+		Client:                 s.cfg.UserClient,
+		Listener:               list,
+		Protocol:               databaseProtocol,
+		WebProxyAddr:           s.webProxyAddr,
+		ProxyMiddleware:        proxyMiddleware,
+		ConnectionDiagnosticID: connectionDiagnosticID,
+		RouteToDatabase: proto.RouteToDatabase{
+			Protocol:    databaseProtocol,
+			Username:    req.DatabaseUser,
+			Database:    req.DatabaseName,
+			ServiceName: databaseServiceName,
+		},
+		Insecure: true, // TODO(marco): remove this
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	pingRequest, err := newPingRequest(list.Addr().String(), req.DatabaseUser, req.DatabaseName)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := databasePinger.Ping(ctx, pingRequest); err != nil {
+		return s.handleErrFromPing(ctx, connectionDiagnosticID, err, databasePinger)
+	}
+
+	return s.handlePingSuccess(ctx, connectionDiagnosticID, databasePinger)
+}
+
+func newPingRequest(alpnProxyAddr, databaseUser, databaseName string) (database.PingRequest, error) {
+	proxyHost, proxyPortStr, err := net.SplitHostPort(alpnProxyAddr)
+	if err != nil {
+		return database.PingRequest{}, trace.Wrap(err)
+	}
+
+	proxyPort, err := strconv.Atoi(proxyPortStr)
+	if err != nil {
+		return database.PingRequest{}, trace.Wrap(err)
+	}
+
+	return database.PingRequest{
+		Host:     proxyHost,
+		Port:     proxyPort,
+		Username: databaseUser,
+		Database: databaseName,
+	}, nil
+}
+
+func (s DatabaseConnectionTester) handlePingSuccess(ctx context.Context, connectionDiagnosticID string, databasePinger DatabasePinger) (types.ConnectionDiagnostic, error) {
```

Comment:
The `databasePinger DatabasePinger` param is not used:
```suggestion
func (s DatabaseConnectionTester) handlePingSuccess(ctx context.Context, connectionDiagnosticID string) (types.ConnectionDiagnostic, error) {
```